File: code/packages/swift/der-asn1/Sources/DerAsn1/DerAsn1.swift
import DerTlv

/// Bounded typed ASN.1 DER values over payload-blind DER framing.
public enum DerAsn1 {
  public typealias Asn1Limits = Limits
  public typealias Asn1ErrorKind = ErrorKind
  public typealias Asn1Error = ValueError
  public typealias Asn1Element = Element
  public typealias Asn1Decoder = Decoder
  public typealias Asn1Cursor = Cursor

  public static let defaultMaxDepth = 32
  public static let defaultMaxTotalElements: UInt64 = 16_384
  public static let defaultMaxOidArcs = 128

  public struct Limits: Sendable {
    public let der: DerTlv.Limits
    public let maxDepth: Int
    public let maxTotalElements: UInt64
    public let maxOidArcs: Int

    public init(
      der: DerTlv.Limits = .defaults,
      maxDepth: Int = DerAsn1.defaultMaxDepth,
      maxTotalElements: UInt64 = DerAsn1.defaultMaxTotalElements,
      maxOidArcs: Int = DerAsn1.defaultMaxOidArcs
    ) {
      precondition(maxDepth >= 0 && maxOidArcs >= 0, "ASN.1 limits must be non-negative")
      self.der = der
      self.maxDepth = maxDepth
      self.maxTotalElements = maxTotalElements
      self.maxOidArcs = maxOidArcs
    }

    fileprivate func matches(_ other: Limits) -> Bool {
      maxDepth == other.maxDepth && maxTotalElements == other.maxTotalElements
        && maxOidArcs == other.maxOidArcs
        && der.maxInputLength == other.der.maxInputLength
        && der.maxValueLength == other.der.maxValueLength
        && der.maxElements == other.der.maxElements
        && der.maxTagNumber == other.der.maxTagNumber
    }
  }

  public enum ErrorKind: String, CaseIterable, Sendable {
    case framing
    case unexpectedTag = "unexpected-tag"
    case decoderLimitMismatch = "decoder-limit-mismatch"
    case depthLimitExceeded = "depth-limit-exceeded"
    case elementLimitExceeded = "element-limit-exceeded"
    case invalidBooleanLength = "invalid-boolean-length"
    case invalidBooleanValue = "invalid-boolean-value"
    case emptyInteger = "empty-integer"
    case nonMinimalInteger = "non-minimal-integer"
    case negativeInteger = "negative-integer"
    case integerOverflow = "integer-overflow"
    case missingUnusedBitCount = "missing-unused-bit-count"
    case invalidUnusedBitCount = "invalid-unused-bit-count"
    case nonZeroBitPadding = "non-zero-bit-padding"
    case bitLengthOverflow = "bit-length-overflow"
    case nonEmptyNull = "non-empty-null"
    case nonAsciiIA5String = "non-ascii-ia5-string"
    case emptyObjectIdentifier = "empty-object-identifier"
    case unterminatedObjectIdentifier = "unterminated-object-identifier"
    case nonMinimalObjectIdentifier = "non-minimal-object-identifier"
    case objectIdentifierOverflow = "object-identifier-overflow"
    case oidArcLimitExceeded = "oid-arc-limit-exceeded"
  }

  public struct ValueError: Error, Equatable, CustomStringConvertible, Sendable {
    public let kind: ErrorKind
    public let offset: Int
    public let framingKind: String?

    public init(kind: ErrorKind, offset: Int, framingKind: String? = nil) {
      self.kind = kind
      self.offset = offset
      self.framingKind = framingKind
    }

    public var description: String {
      "ASN.1 DER value error \(kind.rawValue) at byte \(offset)"
    }
  }

  public struct Element: Sendable {
    public let tag: DerTlv.Tag
    public let depth: Int
    public let header: [UInt8]
    public let value: [UInt8]
    public let encoded: [UInt8]

    fileprivate init(_ element: DerTlv.Element, depth: Int) {
      tag = element.tag
      self.depth = depth
      header = Array(element.header)
      value = Array(element.value)
      encoded = Array(element.encoded)
    }

    fileprivate var valueOffset: Int { header.count }
  }

  public final class Decoder {
    public let limits: Limits
    public private(set) var elementsRead: UInt64 = 0

    public init(limits: Limits = Limits()) { self.limits = limits }

    public func decodeExact(_ input: [UInt8]) throws -> Element {
      guard limits.maxDepth > 0 else { throw failure(.depthLimitExceeded, 0) }
      try requireCapacity(offset: 0)
      do {
        let element = try DerTlv.decodeExact(input, limits: limits.der)
        elementsRead += 1
        return Element(element, depth: 0)
      } catch let error as DerTlv.FramingError {
        throw framing(error)
      }
    }

    public func sequence(_ element: Element) throws -> Cursor {
      try constructed(element, tagClass: "universal", number: 16)
    }

    public func set(_ element: Element) throws -> Cursor {
      try constructed(element, tagClass: "universal", number: 17)
    }

    public func explicit(_ element: Element, tagNumber: UInt64) throws -> Element {
      try expectTag(element, tagClass: "context-specific", constructed: true, number: tagNumber)
      let depth = try childDepth(element)
      try requireCapacity(offset: element.valueOffset)
      do {
        let child = try DerTlv.decodeExact(element.value, limits: limits.der)
        elementsRead += 1
        return Element(child, depth: depth)
      } catch let error as DerTlv.FramingError {
        throw framing(error)
      }
    }

    private func constructed(_ element: Element, tagClass: String, number: UInt64) throws -> Cursor
    {
      try expectTag(element, tagClass: tagClass, constructed: true, number: number)
      let depth = try childDepth(element)
      do {
        return Cursor(
          cursor: try DerTlv.Cursor(element.value, limits: limits.der),
          childDepth: depth,
          limits: limits,
          owner: self
        )
      } catch let error as DerTlv.FramingError {
        throw framing(error)
      }
    }

    private func childDepth(_ element: Element) throws -> Int {
      let depth = element.depth + 1
      guard depth < limits.maxDepth else { throw failure(.depthLimitExceeded, 0) }
      return depth
    }

    fileprivate func requireCapacity(offset: Int) throws {
      guard elementsRead < limits.maxTotalElements else {
        throw failure(.elementLimitExceeded, offset)
      }
    }

    fileprivate func countElement() { elementsRead += 1 }
  }

  public final class Cursor {
    private let cursor: DerTlv.Cursor
    private let childDepth: Int
    private let limits: Limits
    private let owner: Decoder

    fileprivate init(cursor: DerTlv.Cursor, childDepth: Int, limits: Limits, owner: Decoder) {
      self.cursor = cursor
      self.childDepth = childDepth
      self.limits = limits
      self.owner = owner
    }

    public var remaining: [UInt8] { Array(cursor.remaining) }

    public func read(_ decoder: Decoder) throws -> Element? {
      if cursor.remaining.isEmpty { return nil }
      guard decoder === owner && decoder.limits.matches(limits) else {
        throw failure(.decoderLimitMismatch, 0)
      }
      try decoder.requireCapacity(offset: 0)
      do {
        guard let element = try cursor.read() else { return nil }
        decoder.countElement()
        return Element(element, depth: childDepth)
      } catch let error as DerTlv.FramingError {
        throw framing(error)
      }
    }

    public func finish() throws {
      do { try cursor.finish() } catch let error as DerTlv.FramingError { throw framing(error) }
    }
  }

  public struct Integer: Sendable {
    public let signedBytes: [UInt8]
    private let valueOffset: Int
    public var isNegative: Bool { signedBytes[0] & 0x80 != 0 }

    fileprivate init(_ bytes: [UInt8], valueOffset: Int) {
      signedBytes = bytes
      self.valueOffset = valueOffset
    }

    public func toUInt64() throws -> UInt64 {
      guard !isNegative else { throw failure(.negativeInteger, valueOffset) }
      let start = signedBytes[0] == 0 ? 1 : 0
      guard signedBytes.count - start <= 8 else { throw failure(.integerOverflow, valueOffset) }
      var result: UInt64 = 0
      for byte in signedBytes[start...] { result = result * 256 + UInt64(byte) }
      return result
    }
  }

  public struct BitString: Sendable {
    public let bytes: [UInt8]
    public let unusedBits: Int
    public let bitLength: UInt64
  }

  public struct ObjectIdentifier: Sendable {
    public let encoded: [UInt8]
    public let arcs: [UInt64]
    public var arcCount: Int { arcs.count }
    public func equals(arcs expected: [UInt64]) -> Bool { arcs == expected }
  }

  public static func decodeBoolean(_ element: Element) throws -> Bool {
    try expectUniversalPrimitive(element, number: 1)
    guard element.value.count == 1 else {
      throw failure(.invalidBooleanLength, element.valueOffset)
    }
    switch element.value[0] {
    case 0: return false
    case 0xff: return true
    default: throw failure(.invalidBooleanValue, element.valueOffset)
    }
  }

  public static func decodeInteger(_ element: Element) throws -> Integer {
    try expectUniversalPrimitive(element, number: 2)
    guard !element.value.isEmpty else { throw failure(.emptyInteger, element.valueOffset) }
    if element.value.count > 1 {
      let first = element.value[0]
      let second = element.value[1]
      if (first == 0 && second & 0x80 == 0) || (first == 0xff && second & 0x80 != 0) {
        throw failure(.nonMinimalInteger, element.valueOffset)
      }
    }
    return Integer(element.value, valueOffset: element.valueOffset)
  }

  public static func decodeBitString(_ element: Element) throws -> BitString {
    try expectUniversalPrimitive(element, number: 3)
    guard let first = element.value.first else {
      throw failure(.missingUnusedBitCount, element.valueOffset)
    }
    let unused = Int(first)
    let payload = Array(element.value.dropFirst())
    guard unused <= 7 && (!payload.isEmpty || unused == 0) else {
      throw failure(.invalidUnusedBitCount, element.valueOffset)
    }
    if unused != 0, let last = payload.last, last & UInt8((1 << unused) - 1) != 0 {
      throw failure(.nonZeroBitPadding, element.valueOffset + element.value.count - 1)
    }
    let (bits, overflow) = UInt64(payload.count).multipliedReportingOverflow(by: 8)
    guard !overflow else { throw failure(.bitLengthOverflow, element.valueOffset) }
    return BitString(bytes: payload, unusedBits: unused, bitLength: bits - UInt64(unused))
  }

  public static func decodeOctetString(_ element: Element) throws -> [UInt8] {
    try expectUniversalPrimitive(element, number: 4)
    return element.value
  }

  public static func decodeImplicitOctetString(_ element: Element, tagNumber: UInt64) throws
    -> [UInt8]
  {
    try expectContextPrimitive(element, number: tagNumber)
    return element.value
  }

  public static func decodeIA5String(_ element: Element) throws -> String {
    try expectUniversalPrimitive(element, number: 22)
    return try decodeAscii(element)
  }

  public static func decodeImplicitIA5String(_ element: Element, tagNumber: UInt64) throws -> String
  {
    try expectContextPrimitive(element, number: tagNumber)
    return try decodeAscii(element)
  }

  public static func decodeNull(_ element: Element) throws {
    try expectUniversalPrimitive(element, number: 5)
    guard element.value.isEmpty else { throw failure(.nonEmptyNull, element.valueOffset) }
  }

  public static func decodeObjectIdentifier(
    _ element: Element, limits: Limits = Limits()
  ) throws -> ObjectIdentifier {
    try expectUniversalPrimitive(element, number: 6)
    return try decodeOid(element, limits: limits)
  }

  public static func decodeImplicitObjectIdentifier(
    _ element: Element, tagNumber: UInt64, limits: Limits = Limits()
  ) throws -> ObjectIdentifier {
    try expectContextPrimitive(element, number: tagNumber)
    return try decodeOid(element, limits: limits)
  }

  private static func decodeAscii(_ element: Element) throws -> String {
    for (index, byte) in element.value.enumerated() where byte > 0x7f {
      throw failure(.nonAsciiIA5String, element.valueOffset + index)
    }
    return String(decoding: element.value, as: UTF8.self)
  }

  private static func decodeOid(_ element: Element, limits: Limits) throws -> ObjectIdentifier {
    let encoded = element.value
    guard !encoded.isEmpty else { throw failure(.emptyObjectIdentifier, element.valueOffset) }
    let first = try parseArc(encoded, start: 0, valueOffset: element.valueOffset)
    var arcs: [UInt64]
    if first.value < 40 {
      arcs = [0, first.value]
    } else if first.value < 80 {
      arcs = [1, first.value - 40]
    } else {
      arcs = [2, first.value - 80]
    }
    guard arcs.count <= limits.maxOidArcs else {
      throw failure(.oidArcLimitExceeded, element.valueOffset)
    }
    var offset = first.nextOffset
    while offset < encoded.count {
      let start = offset
      let parsed = try parseArc(encoded, start: offset, valueOffset: element.valueOffset)
      arcs.append(parsed.value)
      guard arcs.count <= limits.maxOidArcs else {
        throw failure(.oidArcLimitExceeded, element.valueOffset + start)
      }
      offset = parsed.nextOffset
    }
    return ObjectIdentifier(encoded: encoded, arcs: arcs)
  }

  private static func parseArc(
    _ encoded: [UInt8], start: Int, valueOffset: Int
  ) throws -> (value: UInt64, nextOffset: Int) {
    guard encoded[start] != 0x80 else {
      throw failure(.nonMinimalObjectIdentifier, valueOffset + start)
    }
    var value: UInt64 = 0
    var offset = start
    while true {
      guard offset < encoded.count else {
        throw failure(.unterminatedObjectIdentifier, valueOffset + offset)
      }
      let octet = encoded[offset]
      let payload = UInt64(octet & 0x7f)
      guard value <= (UInt64.max - payload) / 128 else {
        throw failure(.objectIdentifierOverflow, valueOffset + offset)
      }
      value = value * 128 + payload
      offset += 1
      if octet & 0x80 == 0 { return (value, offset) }
    }
  }

  private static func expectUniversalPrimitive(_ element: Element, number: UInt64) throws {
    try expectTag(element, tagClass: "universal", constructed: false, number: number)
  }

  private static func expectContextPrimitive(_ element: Element, number: UInt64) throws {
    try expectTag(element, tagClass: "context-specific", constructed: false, number: number)
  }

  private static func expectTag(
    _ element: Element, tagClass: String, constructed: Bool, number: UInt64
  ) throws {
    guard
      element.tag.tagClass == tagClass && element.tag.constructed == constructed
        && element.tag.number == number
    else { throw failure(.unexpectedTag, 0) }
  }

  private static func failure(_ kind: ErrorKind, _ offset: Int) -> ValueError {
    ValueError(kind: kind, offset: offset)
  }

  private static func framing(_ error: DerTlv.FramingError) -> ValueError {
    ValueError(kind: .framing, offset: error.offset, framingKind: error.kind)
  }
}

File: code/packages/haskell/der-asn1/src/CodingAdventures/DerAsn1.hs
module CodingAdventures.DerAsn1
  ( Asn1Limits (..)
  , defaultAsn1Limits
  , Asn1ErrorKind (..)
  , Asn1Error
  , asn1ErrorKind
  , asn1ErrorOffset
  , asn1ErrorId
  , Asn1Decoder
  , newAsn1Decoder
  , decoderLimits
  , decoderElementsRead
  , decodeExact
  , Asn1Element
  , asn1ElementTag
  , asn1ElementHeader
  , asn1ElementValue
  , asn1ElementEncoded
  , asn1ElementDepth
  , Asn1Cursor
  , sequenceCursor
  , setCursor
  , readAsn1Cursor
  , finishAsn1Cursor
  , asn1CursorRemaining
  , decodeExplicit
  , decodeBoolean
  , DerInteger
  , decodeInteger
  , integerSignedBytes
  , integerIsNegative
  , integerToWord64
  , DerBitString
  , decodeBitString
  , bitStringBytes
  , bitStringUnusedBits
  , bitStringLength
  , decodeOctetString
  , decodeImplicitOctetString
  , decodeIa5String
  , decodeImplicitIa5String
  , decodeNull
  , ObjectIdentifier
  , decodeObjectIdentifier
  , decodeImplicitObjectIdentifier
  , oidEncoded
  , oidArcs
  , oidArcCount
  , oidEquals
  , version
  ) where

import Data.Bits ((.&.), shiftL)
import qualified Data.ByteString as BS
import Data.Char (chr)
import Control.Concurrent.MVar (MVar, modifyMVarMasked, newMVar, readMVar)
import qualified Data.Text as T
import Data.Word (Word32, Word64, Word8)
import qualified CodingAdventures.DerTlv as DT

version :: String
version = "0.1.0"

data Asn1Limits = Asn1Limits
  { asn1DerLimits :: DT.DerLimits
  , maxDepth :: Word64
  , maxTotalElements :: Word64
  , maxOidArcs :: Word64
  }
  deriving (Eq)

defaultAsn1Limits :: Asn1Limits
defaultAsn1Limits = Asn1Limits DT.defaultDerLimits 32 16384 128

data Asn1ErrorKind
  = Framing DT.DerErrorKind
  | UnexpectedTag
  | DecoderLimitMismatch
  | DepthLimitExceeded
  | ElementLimitExceeded
  | InvalidBooleanLength
  | InvalidBooleanValue
  | EmptyInteger
  | NonMinimalInteger
  | NegativeInteger
  | IntegerOverflow
  | MissingUnusedBitCount
  | InvalidUnusedBitCount
  | NonZeroBitPadding
  | BitLengthOverflow
  | NonEmptyNull
  | NonAsciiIa5String
  | EmptyObjectIdentifier
  | UnterminatedObjectIdentifier
  | NonMinimalObjectIdentifier
  | ObjectIdentifierOverflow
  | OidArcLimitExceeded
  deriving (Eq, Show)

data Asn1Error = Asn1Error
  { asn1ErrorKind :: Asn1ErrorKind
  , asn1ErrorOffset :: Int
  }

instance Show Asn1Error where
  show err = "ASN.1 DER value error " ++ show (asn1ErrorKind err)
    ++ " at byte " ++ show (asn1ErrorOffset err)

asn1ErrorId :: Asn1ErrorKind -> String
asn1ErrorId kind = case kind of
  Framing _ -> "framing"
  UnexpectedTag -> "unexpected-tag"
  DecoderLimitMismatch -> "decoder-limit-mismatch"
  DepthLimitExceeded -> "depth-limit-exceeded"
  ElementLimitExceeded -> "element-limit-exceeded"
  InvalidBooleanLength -> "invalid-boolean-length"
  InvalidBooleanValue -> "invalid-boolean-value"
  EmptyInteger -> "empty-integer"
  NonMinimalInteger -> "non-minimal-integer"
  NegativeInteger -> "negative-integer"
  IntegerOverflow -> "integer-overflow"
  MissingUnusedBitCount -> "missing-unused-bit-count"
  InvalidUnusedBitCount -> "invalid-unused-bit-count"
  NonZeroBitPadding -> "non-zero-bit-padding"
  BitLengthOverflow -> "bit-length-overflow"
  NonEmptyNull -> "non-empty-null"
  NonAsciiIa5String -> "non-ascii-ia5-string"
  EmptyObjectIdentifier -> "empty-object-identifier"
  UnterminatedObjectIdentifier -> "unterminated-object-identifier"
  NonMinimalObjectIdentifier -> "non-minimal-object-identifier"
  ObjectIdentifierOverflow -> "object-identifier-overflow"
  OidArcLimitExceeded -> "oid-arc-limit-exceeded"

framing :: DT.DerError -> Asn1Error
framing errorValue = Asn1Error (Framing (DT.derErrorKind errorValue)) (DT.derErrorOffset errorValue)

data Asn1Decoder = Asn1Decoder Asn1Limits (MVar Word64)

newAsn1Decoder :: Asn1Limits -> IO Asn1Decoder
newAsn1Decoder limits = Asn1Decoder limits <$> newMVar 0

decoderLimits :: Asn1Decoder -> Asn1Limits
decoderLimits (Asn1Decoder limits _) = limits

decoderElementsRead :: Asn1Decoder -> IO Word64
decoderElementsRead (Asn1Decoder _ counter) = readMVar counter

data Asn1Element = Asn1Element DT.DerElement Word64 (MVar Word64)

asn1ElementTag :: Asn1Element -> DT.DerTag
asn1ElementTag (Asn1Element element _ _) = DT.elementTag element

asn1ElementHeader :: Asn1Element -> BS.ByteString
asn1ElementHeader (Asn1Element element _ _) = DT.elementHeader element

asn1ElementValue :: Asn1Element -> BS.ByteString
asn1ElementValue (Asn1Element element _ _) = DT.elementValue element

asn1ElementEncoded :: Asn1Element -> BS.ByteString
asn1ElementEncoded (Asn1Element element _ _) = DT.elementEncoded element

asn1ElementDepth :: Asn1Element -> Word64
asn1ElementDepth (Asn1Element _ depth _) = depth

valueOffset :: Asn1Element -> Int
valueOffset = BS.length . asn1ElementHeader

requireOwner :: Asn1Decoder -> MVar Word64 -> Either Asn1Error ()
requireOwner (Asn1Decoder _ owner) expected
  | owner == expected = Right ()
  | otherwise = Left (Asn1Error DecoderLimitMismatch 0)

decodeExact :: Asn1Decoder -> BS.ByteString -> IO (Either Asn1Error Asn1Element)
decodeExact (Asn1Decoder limits owner) input
  | maxDepth limits == 0 = pure (Left (Asn1Error DepthLimitExceeded 0))
  | otherwise = modifyMVarMasked owner $ \count ->
      if count >= maxTotalElements limits
        then pure (count, Left (Asn1Error ElementLimitExceeded 0))
        else case DT.decodeExact (asn1DerLimits limits) input of
          Left err -> pure (count, Left (framing err))
          Right element -> pure (count + 1, Right (Asn1Element element 0 owner))

data Asn1Cursor = Asn1Cursor DT.DerCursor Word64 Asn1Limits (MVar Word64)

childDepth :: Asn1Limits -> Asn1Element -> Either Asn1Error Word64
childDepth limits element =
  let depth = asn1ElementDepth element + 1
  in if depth >= maxDepth limits
       then Left (Asn1Error DepthLimitExceeded 0)
       else Right depth

expectTag :: DT.TagClass -> Bool -> Word32 -> Asn1Element -> Either Asn1Error ()
expectTag expectedClass expectedConstructed expectedNumber element =
  let tag = asn1ElementTag element
  in if DT.tagClass tag == expectedClass
        && DT.tagConstructed tag == expectedConstructed
        && DT.tagNumber tag == expectedNumber
       then Right ()
       else Left (Asn1Error UnexpectedTag 0)

constructedCursor :: DT.TagClass -> Word32 -> Asn1Decoder -> Asn1Element -> Either Asn1Error Asn1Cursor
constructedCursor expectedClass expectedNumber decoder@(Asn1Decoder limits owner) element@(Asn1Element _ _ elementOwner) = do
  requireOwner decoder elementOwner
  expectTag expectedClass True expectedNumber element
  depth <- childDepth limits element
  cursor <- either (Left . framing) Right (DT.newCursor (asn1DerLimits limits) (asn1ElementValue element))
  pure (Asn1Cursor cursor depth limits owner)

sequenceCursor :: Asn1Decoder -> Asn1Element -> Either Asn1Error Asn1Cursor
sequenceCursor = constructedCursor DT.Universal 16

setCursor :: Asn1Decoder -> Asn1Element -> Either Asn1Error Asn1Cursor
setCursor = constructedCursor DT.Universal 17

readAsn1Cursor :: Asn1Decoder -> Asn1Cursor -> IO (Either Asn1Error (Maybe Asn1Element, Asn1Cursor))
readAsn1Cursor (Asn1Decoder limits owner) cursor@(Asn1Cursor lower depth cursorLimits cursorOwner)
  | BS.null (DT.cursorRemaining lower) = pure (Right (Nothing, cursor))
  | limits /= cursorLimits || owner /= cursorOwner = pure (Left (Asn1Error DecoderLimitMismatch 0))
  | otherwise = modifyMVarMasked owner $ \count ->
      if count >= maxTotalElements limits
        then pure (count, Left (Asn1Error ElementLimitExceeded 0))
        else case DT.readCursor lower of
          Left err -> pure (count, Left (framing err))
          Right (Nothing, next) -> pure (count, Right (Nothing, Asn1Cursor next depth cursorLimits cursorOwner))
          Right (Just element, next) ->
            pure (count + 1, Right (Just (Asn1Element element depth owner), Asn1Cursor next depth cursorLimits cursorOwner))

finishAsn1Cursor :: Asn1Cursor -> Either Asn1Error ()
finishAsn1Cursor (Asn1Cursor cursor _ _ _) = either (Left . framing) Right (DT.finishCursor cursor)

asn1CursorRemaining :: Asn1Cursor -> BS.ByteString
asn1CursorRemaining (Asn1Cursor cursor _ _ _) = DT.cursorRemaining cursor

decodeExplicit :: Asn1Decoder -> Asn1Element -> Word32 -> IO (Either Asn1Error Asn1Element)
decodeExplicit decoder@(Asn1Decoder limits owner) element@(Asn1Element _ _ elementOwner) tagNumber =
  case requireOwner decoder elementOwner >> expectTag DT.ContextSpecific True tagNumber element >> childDepth limits element of
    Left err -> pure (Left err)
    Right depth -> modifyMVarMasked owner $ \count ->
      if count >= maxTotalElements limits
        then pure (count, Left (Asn1Error ElementLimitExceeded (valueOffset element)))
        else case DT.decodeExact (asn1DerLimits limits) (asn1ElementValue element) of
          Left err -> pure (count, Left (framing err))
          Right child -> pure (count + 1, Right (Asn1Element child depth owner))

expectUniversalPrimitive :: Word32 -> Asn1Element -> Either Asn1Error ()
expectUniversalPrimitive number = expectTag DT.Universal False number

expectContextPrimitive :: Word32 -> Asn1Element -> Either Asn1Error ()
expectContextPrimitive number = expectTag DT.ContextSpecific False number

decodeBoolean :: Asn1Element -> Either Asn1Error Bool
decodeBoolean element = do
  expectUniversalPrimitive 1 element
  case BS.unpack (asn1ElementValue element) of
    [0] -> Right False
    [255] -> Right True
    [_] -> Left (Asn1Error InvalidBooleanValue (valueOffset element))
    _ -> Left (Asn1Error InvalidBooleanLength (valueOffset element))

data DerInteger = DerInteger BS.ByteString Int

integerSignedBytes :: DerInteger -> BS.ByteString
integerSignedBytes (DerInteger bytes _) = bytes

integerIsNegative :: DerInteger -> Bool
integerIsNegative (DerInteger bytes _) = BS.head bytes .&. 0x80 /= 0

integerToWord64 :: DerInteger -> Either Asn1Error Word64
integerToWord64 integer@(DerInteger bytes offset)
  | integerIsNegative integer = Left (Asn1Error NegativeInteger offset)
  | BS.length magnitude > 8 = Left (Asn1Error IntegerOverflow offset)
  | otherwise = Right (BS.foldl' (\value byte -> shiftL value 8 + fromIntegral byte) 0 magnitude)
  where
    magnitude = if BS.head bytes == 0 then BS.tail bytes else bytes

decodeInteger :: Asn1Element -> Either Asn1Error DerInteger
decodeInteger element = do
  expectUniversalPrimitive 2 element
  let bytes = asn1ElementValue element
      offset = valueOffset element
  if BS.null bytes
    then Left (Asn1Error EmptyInteger offset)
    else if BS.length bytes > 1
      && ((BS.index bytes 0 == 0 && BS.index bytes 1 .&. 0x80 == 0)
        || (BS.index bytes 0 == 255 && BS.index bytes 1 .&. 0x80 /= 0))
      then Left (Asn1Error NonMinimalInteger offset)
      else Right (DerInteger bytes offset)

data DerBitString = DerBitString BS.ByteString Word8 Word64

bitStringBytes :: DerBitString -> BS.ByteString
bitStringBytes (DerBitString bytes _ _) = bytes

bitStringUnusedBits :: DerBitString -> Word8
bitStringUnusedBits (DerBitString _ unused _) = unused

bitStringLength :: DerBitString -> Word64
bitStringLength (DerBitString _ _ lengthValue) = lengthValue

decodeBitString :: Asn1Element -> Either Asn1Error DerBitString
decodeBitString element = do
  expectUniversalPrimitive 3 element
  let value = asn1ElementValue element
      offset = valueOffset element
  if BS.null value then Left (Asn1Error MissingUnusedBitCount offset) else do
    let unused = BS.head value
        bytes = BS.tail value
    if unused > 7 || (BS.null bytes && unused /= 0)
      then Left (Asn1Error InvalidUnusedBitCount offset)
      else if unused /= 0 && BS.last bytes .&. (shiftL 1 (fromIntegral unused) - 1) /= 0
        then Left (Asn1Error NonZeroBitPadding (offset + BS.length value - 1))
        else
          let byteCount = fromIntegral (BS.length bytes) :: Word64
          in if byteCount > maxBound `div` 8
               then Left (Asn1Error BitLengthOverflow offset)
               else Right (DerBitString bytes unused (byteCount * 8 - fromIntegral unused))

decodeOctetString :: Asn1Element -> Either Asn1Error BS.ByteString
decodeOctetString element = expectUniversalPrimitive 4 element >> Right (asn1ElementValue element)

decodeImplicitOctetString :: Asn1Element -> Word32 -> Either Asn1Error BS.ByteString
decodeImplicitOctetString element tagNumber = expectContextPrimitive tagNumber element >> Right (asn1ElementValue element)

decodeIa5Contents :: Asn1Element -> Either Asn1Error T.Text
decodeIa5Contents element =
  case BS.findIndex (> 127) bytes of
    Just index -> Left (Asn1Error NonAsciiIa5String (valueOffset element + index))
    Nothing -> Right (T.pack (map (chr . fromIntegral) (BS.unpack bytes)))
  where bytes = asn1ElementValue element

decodeIa5String :: Asn1Element -> Either Asn1Error T.Text
decodeIa5String element = expectUniversalPrimitive 22 element >> decodeIa5Contents element

decodeImplicitIa5String :: Asn1Element -> Word32 -> Either Asn1Error T.Text
decodeImplicitIa5String element tagNumber = expectContextPrimitive tagNumber element >> decodeIa5Contents element

decodeNull :: Asn1Element -> Either Asn1Error ()
decodeNull element = do
  expectUniversalPrimitive 5 element
  if BS.null (asn1ElementValue element)
    then Right ()
    else Left (Asn1Error NonEmptyNull (valueOffset element))

data ObjectIdentifier = ObjectIdentifier BS.ByteString [Word64]

oidEncoded :: ObjectIdentifier -> BS.ByteString
oidEncoded (ObjectIdentifier bytes _) = bytes

oidArcs :: ObjectIdentifier -> [Word64]
oidArcs (ObjectIdentifier _ arcs) = arcs

oidArcCount :: ObjectIdentifier -> Word64
oidArcCount = fromIntegral . length . oidArcs

oidEquals :: ObjectIdentifier -> [Word64] -> Bool
oidEquals oid expected = oidArcs oid == expected

parseBase128 :: BS.ByteString -> Int -> Int -> Either Asn1Error (Word64, Int)
parseBase128 bytes start baseOffset
  | start >= BS.length bytes = Left (Asn1Error UnterminatedObjectIdentifier (baseOffset + start))
  | BS.index bytes start == 0x80 = Left (Asn1Error NonMinimalObjectIdentifier (baseOffset + start))
  | otherwise = go start 0
  where
    go index value
      | index >= BS.length bytes = Left (Asn1Error UnterminatedObjectIdentifier (baseOffset + index))
      | value > (maxBound - fromIntegral payload) `div` 128 = Left (Asn1Error ObjectIdentifierOverflow (baseOffset + index))
      | octet .&. 0x80 == 0 = Right (candidate, index + 1)
      | otherwise = go (index + 1) candidate
      where
        octet = BS.index bytes index
        payload = octet .&. 0x7f
        candidate = value * 128 + fromIntegral payload

decodeOidContents :: Asn1Limits -> Asn1Element -> Either Asn1Error ObjectIdentifier
decodeOidContents limits element = do
  let bytes = asn1ElementValue element
      base = valueOffset element
  if BS.null bytes then Left (Asn1Error EmptyObjectIdentifier base) else do
    (combined, firstEnd) <- parseBase128 bytes 0 base
    let prefix = if combined < 40 then [0, combined]
          else if combined < 80 then [1, combined - 40]
          else [2, combined - 80]
    if maxOidArcs limits < 2
      then Left (Asn1Error OidArcLimitExceeded base)
      else do
        arcs <- parseRest bytes base firstEnd prefix
        Right (ObjectIdentifier bytes arcs)
  where
    parseRest bytes base index arcs
      | index >= BS.length bytes = Right arcs
      | otherwise = do
          (arc, next) <- parseBase128 bytes index base
          let nextArcs = arcs ++ [arc]
          if fromIntegral (length nextArcs) > maxOidArcs limits
            then Left (Asn1Error OidArcLimitExceeded (base + index))
            else parseRest bytes base next nextArcs

decodeObjectIdentifier :: Asn1Element -> Asn1Limits -> Either Asn1Error ObjectIdentifier
decodeObjectIdentifier element limits = expectUniversalPrimitive 6 element >> decodeOidContents limits element

decodeImplicitObjectIdentifier :: Asn1Element -> Word32 -> Asn1Limits -> Either Asn1Error ObjectIdentifier
decodeImplicitObjectIdentifier element tagNumber limits = expectContextPrimitive tagNumber element >> decodeOidContents limits element
